laurent22/joplin
# Mobile: Log screen: Add share menu item

# Summary
This pull request
1. Converts `log.js` to TypeScript (as per [coding_style.md](https://github.com/laurent22/joplin/blob/dev/readme/coding_style.md#convert-existing-js-files-to-typescript-before-modifying))
2. Adds a "share" menu item to the log screen.

See also [the relevant forum post](https://discourse.joplinapp.org/t/feature-request-copy-logs-button-on-mobile/31411).

# Testing
As the log screen is a connected component, it presently does not have automated tests.

<details>
<summary>Issues adding automated tests</summary>

`jest` has trouble with some of the files imported by dependencies of `LogScreen.tsx`. Although updating [`transformIgnorePatterns`](https://jestjs.io/docs/tutorial-react-native#transformignorepatterns-customization) to transform files in `node_modules` fixed some of these issues, it fails to import a module in `highlight.js`:


```
Cannot find module '@babel/runtime/helpers/interopRequireDefault' from '../renderer/node_modules/highlight.js/lib/core.js'

    Require stack:
      /Users/Henry/Documents/Build/joplin/packages/renderer/node_modules/highlight.js/lib/core.js
      /Users/Henry/Documents/Build/joplin/packages/renderer/highlight.ts
      /Users/Henry/Documents/Build/joplin/packages/renderer/MdToHtml.ts
      /Users/Henry/Documents/Build/joplin/packages/renderer/MarkupToHtml.ts
      /Users/Henry/Documents/Build/joplin/packages/renderer/index.js
      /Users/Henry/Documents/Build/joplin/packages/lib/markdownUtils.ts
      /Users/Henry/Documents/Build/joplin/packages/lib/models/BaseItem.ts
      /Users/Henry/Documents/Build/joplin/packages/lib/models/Note.ts
      components/ScreenHeader.tsx
      components/screens/LogScreen.tsx
      components/screens/LogScreen.test.ts

       5 | // https://github.com/highlightjs/highlight.js/issues/1257#issuecomment-254502834
       6 |
    >  7 | const hljs = require('highlight.js/lib/core');
         |              ^
       8 |
       9 | // hljs.registerLanguage('1c', require('highlight.js/lib/languages/1c'));
      10 | hljs.registerLanguage('abnf', require('highlight.js/lib/languages/abnf'));

      at Resolver._throwModNotFoundError (node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.<anonymous> (../renderer/node_modules/highlight.js/lib/core.js:1:50)
      at Object.<anonymous> (../renderer/highlight.ts:7:14)
      at Object.<anonymous> (../renderer/MdToHtml.ts:9:1)
      at Object.<anonymous> (../renderer/MarkupToHtml.ts:1:1)
      at Object.require (../renderer/index.ts:1:1)
      at Object.<anonymous> (../lib/markdownUtils.ts:1:1)
      at Object.<anonymous> (../lib/models/BaseItem.ts:6:1)
      at Object.<anonymous> (../lib/models/Note.ts:2:1)
      at Object.<anonymous> (components/ScreenHeader.tsx:12:1)
      at Object.<anonymous> (components/screens/LogScreen.tsx:6:1)
      at Object.<anonymous> (components/screens/LogScreen.test.ts:5:1)

Test Suites: 1 failed, 7 passed, 8 total
Tests:       45 passed, 45 total
Snapshots:   0 total
Time:        12.77 s, estimated 17 s
Ran all test suites.
```

See also [React Redux's testing documentation](https://redux.js.org/usage/writing-tests#connected-components).

</details>

To manually verify that this feature is working correctly,
1. Open the "Log" screen (settings > log)
2. Click "share"
4. Click "errors only"
5. Click "share"

In steps 2 and 4, "share" should only share the text visible in the log.

To manually verify that related features have not broken,
1. Open a note with an image resource attachment
2. Share the attached resource using a long-press gesture.

# To-do
- [x] Test on iOS
- [x] Show error message if insufficient space to store a copy of the log file.
- [x] Add automated test for changes to ShareUtils.ts
